### PR TITLE
Improvements - method implementation pattern and naming convention

### PR DESCRIPTION
**Pull request by chetkhatri**
What were proposed in this Pull Request?
* avoid mutability at the method `findLastCompleteCheckpoint` and written recursive implementation.
* method `override def listFrom(path: Path)` actually returns the `Iterator` but name is `listFrom` which should be `iteratorFrom`. I came across this, when I debug the code
```
  val checkpoints =
          store
            .listFrom(checkpointPrefix(logPath, math.max(0, cur - 1000)))
```
`.iterate` was not resolved. Which remind me the method doesn't return the `List`.

How this PR was tested?
* compiled with sbt - PASSED
* CI / CD run - PASSED

@brkyvz Request for the review.

### CODE REVIEW

**Comment by chetkhatri** [comment]
@brkyvz Ping for you.

**Comment by chetkhatri** [comment]
@brkyvz PING for review.

**Comment by mukulmurthy** [comment]
Sorry for the delay reviewing; Burak is on vacation right now. I don't see a problem with the name listFrom. It's listing files from a Log Store, and it returns an iterator for performance reasons.

I'd approve the change to use recursion instead of mutability in Checkpoints.scala by itself.

**Comment by chetkhatri** [comment]
@mukulmurthy Thanks for review.

**Comment by chetkhatri** [comment]
Let's wait for @brkyvz 's review.

**Comment by chetkhatri** [comment]
@brkyvz PING if you're back.

File: src/main/scala/org/apache/spark/sql/delta/storage/FileSystemLogStore.scala [hunk 1]
```diff
@@ -52,7 +52,7 @@ abstract class FileSystemLogStore(
     }
   }
 
-  override def listFrom(path: Path): Iterator[FileStatus] = {
+  override def iteratorFrom(path: Path): Iterator[FileStatus] = {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/storage/FileSystemLogStore.scala]: Can you please revert this change? The `listFrom` here doesn't correspond to returning a `List`. It's corresponding to a file listing action on a storage system. There's no reason to break a developer API here

File: src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala [hunk 2]
```diff
@@ -168,22 +168,26 @@ trait Checkpoints extends DeltaLogging {
...
-          .map(_.getPath)
-          .filter(isCheckpointFile)
-          .map(CheckpointInstance(_))
-          .takeWhile(tv => (cur == 0 || tv.version <= cur) && tv.isEarlierThan(cv))
-          .toArray
-      val lastCheckpoint = getLatestCompleteCheckpointFromList(checkpoints, cv)
-      if (lastCheckpoint.isDefined) {
-        return lastCheckpoint
+    @annotation.tailrec
+    def loop(cur: Long): Option[CheckpointInstance] =
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala]: I don't think making this call a tail recursion makes this code any more readable. In fact it adds friction to developers who don't know Scala in understanding the code base

File: src/main/scala/org/apache/spark/sql/delta/storage/FileSystemLogStore.scala (hunk 1 above)
Comment (chetkhatri) [reply src/main/scala/org/apache/spark/sql/delta/storage/FileSystemLogStore.scala]: @brkyvz Oh, I see. thanks. I'll revert.

File: src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala (hunk 2 above)
Comment (chetkhatri) [reply src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala]: @brkyvz Sure, In that case, I am closing this PR. Thanks

**Comment by chetkhatri** [comment]
@brkyvz closing this.
